File: 1034-subarrays-with-k-different-integers/1034-subarrays-with-k-different-integers.java
class Solution {
     public int slidingWindow(int[] nums, int k) {
        int n = nums.length;
        int ans = 0;
        Map<Integer,Integer> mp = new HashMap<>();
        int i = 0, j = 0;
        while(j < n){
            mp.put(nums[j], mp.getOrDefault(nums[j],0) + 1);
            while(mp.size() > k){
                mp.put(nums[i], mp.get(nums[i])-1);
                if(mp.get(nums[i]) == 0){
                    mp.remove(nums[i]);
                }
                i++;
            }
            ans += (j-i+1);
            j++;
        }
        return ans;
    }
    public int subarraysWithKDistinct(int[] nums, int k) {
        return slidingWindow(nums, k) - slidingWindow(nums, k-1);
    }
}

File: 1034-subarrays-with-k-different-integers/1034-subarrays-with-k-different-integers.cpp
class Solution {
public:
    int slidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        int ans = 0;
        unordered_map<int,int> mp;
        int i = 0, j = 0;
        while(j < n){
            mp[nums[j]]++;
            while(mp.size() > k){
                mp[nums[i]]--;
                if(mp[nums[i]] == 0){
                    mp.erase(nums[i]);
                }
                i++;
            }
            ans += (j-i+1);
            j++;
        }
        return ans;
    }
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        int subarrWithKDiffEle = slidingWindow(nums,k);
        int subarrWithLessKDiff = slidingWindow(nums, k-1);

        return subarrWithKDiffEle - subarrWithLessKDiff;
    }
};